### DISCUSSION TITLE
InlineData execution order starting from xunit 2.5.1

### DISCUSSION THREAD

**Discussion by mouad-bajjouk** [xunit/xunit#2805] [Question] [answered]
Hi, does anyone get a problem of InlineData execution order starting from xunit 2.5.1
I have a unit test method that rely on the previous test execution in order to make further assertions, the attribute contains a boolean value, and it should start by false, then true:
```
[Theory]
[InlineData(false)]
[InlineData(true)]
```
but when running the test both locally and in my Azure build pipeline, it starts by the `true` value then `false`, which leads to the test failure.
I already tried looking at the open issues and release notes, but I didn't found any help there.
Does anyone have a workaround for this issue?

**Comment by bradwilson** [comment, marked as answer]
This behavior is undefined (and therefore by design). The behavior you've been depending has been accidental.

Additionally, our default test case orderer randomizes tests precisely to discourage order dependency (as that's an anti-pattern in pure unit testing). Looking at the [blame of `DefaultTestCaseOrderer`](https://github.com/xunit/xunit/blame/v2/src/xunit.execution/Sdk/DefaultTestCaseOrderer.cs), nothing has changed there since 2015 (and even then, it was still fundamentally the same as it's been for a very long time). It depends on the [implementation of the unique ID of the test case](https://github.com/xunit/xunit/blame/v2/src/xunit.execution/Sdk/Frameworks/TestMethodTestCase.cs#L215-L249), and that has also been stable for 6+ years.

I can see no reason why 2.5.1 should change the order of what you're seeing, _but that's irrelevant because that order has never been guaranteed_.

If your tests are order dependent, you will need to find a third party extension that allows you to ensure the order that test data is run in. We do not have any facility for this, nor do we have any samples that show it (we have built-in extensibility for ordering test cases and test collections, but not test data rows; as such, any extension here would be pretty involved to write).

My suggestion might be to move the test method logic into a private method, and then instead of having a data-driven test, your test method becomes a `[Fact]`, and then you just invoke the test method twice, once with `false` and once with `true`. You could guarantee both run by using `Assert.Multiple`, once for each invocation, and then if they both fail, you'll see both failure results. (My guess is you don't actually want this, though, because if the invocation of `false` is required to make `true` pass, then if the `false` invocation fails you wouldn't want to run `true` because the prerequisites might not have been fulfilled.)

**Reply by mouad-bajjouk** [reply]
Alright, thanks for the help @bradwilson !

**Comment by Ghostbird** [comment]
The fact that theory data is randomised forcefully is extremely annoying. It means that I cannot properly debug my failing theories, because I have a long list of entries that I need to step through in two different implementations, but I cannot get them to align, due to randomisation.

Note: It's not InlineData specific. I've explicitly created a generator that provides the data in a specific order. But apparently xunit reads the entire generator. Changes the order, then runs the tests.

Also the order is not randomised. The order used is always the same, but _not_ the order in which they are supplied.

**Reply by bradwilson** [reply]
> The fact that theory data is randomised forcefully is extremely annoying. It means that I cannot properly debug my failing theories, because I have a long list of entries that I need to step through in two different implementations, but I cannot get them to align, due to randomisation.

Randomization has been a feature since v1. The intent is to try to catch tests that are unintentionally written to depend on the results of previous tests.

For v1 and v2, if you want tests always to run in a specific order, we offer that as an extensibility point, with example code: https://github.com/xunit/samples.xunit/tree/main/TestOrderExamples. This is the only way to influence the order tests run in for v1 and v2.

For v3, we have implemented a randomization "seed" system that should help in two ways:

- The order will remain stable so long as you don't recompile the test project (as the default seed is based on the [.NET Module Version ID](https://github.com/xunit/xunit/blob/d4060c003440bce08cbc828ec5953435baac8840/src/xunit.v3.core/Sdk/v3/Framework/TestFrameworkExecutor.cs#L82)), so running the tests over and over again will run them in the same order.
- The seed that's used for randomization can be printed when the test is run (by enabling [diagnostic messages](https://xunit.net/docs/configuration-files#diagnosticMessages), and if you're using `dotnet test`, passing `--logger "console;verbosity=normal"` as a command line switch so that xUnit.net's messages show up in the output). A [configuration file item](https://xunit.net/docs/configuration-files#seed) can be used to set the randomization seed so that the issue becomes repeatable even after re-compiling.

It's important to note that _**adding or removing tests from the test project will by definition change the way randomization works**_ even when using a fixed seed, so if you want a permanently stable test order, then use something like the sample linked above.

**Reply by Ghostbird** [reply]
Ah, thanks a lot. The test case ordering option is nice. I managed to debug my issue this time, by manually commenting out all other inline data. I don't need a permanently fixed order, and I get the reason for this functionality. I just sometimes want to disable it when hunting a bug.

In this case I had a function that seemingly failed to produce valid calculations in certain situations. The implementation was transcribed exactly from an older implementation in another language. So I wanted to debug the tests for both versions in lockstep to see where they went out of sync. It turned out to be caused by differences in operator precedence. Part of an expression needed parentheses in C# in order to be equivalent to the original implementation.
